Go attempt

ParamPatel207 opened the pull request:
Updated Readme and required changes.

ParamPatel207 committed:
updated packages

ParamPatel207 committed:
updated lib name

ParamPatel207 committed:
updated comments,separation of redeploy from parsing logic and package path

ParamPatel207 committed:
updated main func

ParamPatel207 committed:
modification for main and mylib completed

ParamPatel207 committed:
all testing completed and code refactored

ParamPatel207 committed:
all testing completed and code refactored

ParamPatel207 committed:
all testing completed and code refactored

ParamPatel207 committed:
all testing completed and code refactored

ParamPatel207 committed:
all testing completed and code refactored

ParamPatel207 committed:
updated readme and comments for the functions

ParamPatel207 committed:
updated readme and comments for the functions

ParamPatel207 committed:
updated readme

ParamPatel207 committed:
updated readme

jonfriesen commented on README.md:
```diff
@@ -1,8 +1,13 @@
+### DigitalOcean App Platform Auto-deploy with same app spec.
+- Get DigitalOcean Personal Access token by following this [instructions](https://docs.digitalocean.com/reference/api/create-personal-access-token/).**(skip this step if you already have DigitalOcean Personal Access Token)**
 - Declare DigitalOcean Personal Access Token as DIGITALOCEAN_ACCESS_TOKEN variable in the [secrets](https://docs.github.com/en/actions/reference/encrypted-secrets#creating-encrypted-secrets-for-a-repository) of github repository. 
 - Add this step to deploy your application on DigitalOcean App Platform without changing any app spec configuration or making any other changes.
```
```suggestion
- [Create a GitHub Action workflow file](https://docs.github.com/en/actions/learn-github-actions/introduction-to-github-actions#create-an-example-workflow) and add this step below to it or add this to your existing action.
```

jonfriesen commented on README.md:
```diff
@@ -1,8 +1,13 @@
 # Usage
-### DigitalOcean App Platform redeploy with same app spec.
-- (skip this step if you already have DigitalOcean Personal Access Token) Get DigitalOcean Personal Access token by following this [instructions](https://docs.digitalocean.com/reference/api/create-personal-access-token/)
+### DigitalOcean App Platform Auto-deploy with same app spec.
```
```suggestion
### Deploy via GH Action and let DigitalOcean App Platform build and deploy your app.
```

jonfriesen commented on README.md:
```diff
@@ -1,8 +1,13 @@
-# DigitalOcean App Platform Image and DigitalOcean Container Registry publish
-This action can be used to redeploy application on the DigitalOcean's [App Platform](https://www.digitalocean.com/products/app-platform/) using github action. This Action has two use cases one is to redeploy your application on App Platform with same configuration. The other use case is to update the DigitalOcean Container Registry configuration and deploy to App Platform. This github action uses ...
+# Deploy a [DigitalOcean App Platform](https://www.digitalocean.com/products/app-platform/) app using GitHub Actions.
```
Do we need a disclaimer that there app has to already exist?

ParamPatel207 replied:
How about?
# Redeploy a [DigitalOcean App Platform](https://www.digitalocean.com/products/app-platform/) app using GitHub Actions.

ParamPatel207 replied:
i think above title can imply that app should be deployed without disclaimer

jonfriesen replied:
I think it'd be confusing. Let's leave it as is and see if it becomes a problem.

jonfriesen commented on README.md:
````diff
@@ -14,10 +19,11 @@ This action can be used to redeploy application on the DigitalOcean's [App Platf
   ```
 - DigitalOcean App Platform will now deploy your application.
````
```suggestion
- This step will trigger a deploy to your App on DigitalOcean App Platform
```

ParamPatel207 replied:
updated using Commit suggestion option.

jonfriesen commented on README.md:
```diff
@@ -14,10 +19,11 @@ This action can be used to redeploy application on the DigitalOcean's [App Platf
 
-### Update DigitalOcean Container Registry of multiple component in App Spec
-- (skip this step if you already have DigitalOcean Personal Access Token) Get DigitalOcean Personal Access token by following this [instructions](https://docs.digitalocean.com/reference/api/create-personal-access-token/)
+### Update DigitalOcean Container Registry(DOCR) configuration of multiple component in App Spec
```
```suggestion
### Deploy an one or more app components from a DigitalOcean Container Registry (DOCR) 
```

ParamPatel207 replied:
updated using commit suggestion

jonfriesen commented on README.md:
```diff
@@ -38,11 +44,12 @@ This action can be used to redeploy application on the DigitalOcean's [App Platf
                       ]'
```
Add some example data

ParamPatel207 replied:
I intentionally skipped the example data so that user can copy the format and paste it in their workflow. Because while doing research for headers i found that while reading documentation many users prefer to copy from documentation prompt.

jonfriesen commented on README.md:
````diff
@@ -38,11 +44,12 @@ This action can be used to redeploy application on the DigitalOcean's [App Platf
                       ]'
   ```
 - DigitalOcean App Platform will now update your DOCR information in App Spec and then deploy your application.
````
```suggestion
- DigitalOcean App Platform will now update your DOCR information in App Spec and then deploy your application.
- This step will trigger a DigitalOcean App Platform deployment of your app using the images specified.
```

jonfriesen commented on README.md:
```diff
@@ -38,11 +44,12 @@ This action can be used to redeploy application on the DigitalOcean's [App Platf
 - DigitalOcean App Platform will now update your DOCR information in App Spec and then deploy your application.
-  (Please use unique tag value for DigitalOcean Container Registry Push instead of latest to update DOCR update in App Spec)
+
+**Note: Always use unique tag names to push image to the DigitalOcean Container Registry. This will allow you to deploy your application without delay. [ref](https://docs.digitalocean.com/products/container-registry/quickstart/)**
```
Give an example, eg using git short hashes

ParamPatel207 replied:
git short hashes is shown in example https://github.com/ParamPatel207/docr_sample. Should I use

- name: Install doctl
  uses: digitalocean/action-doctl@v2
  with:
  token: ${{ secrets.DIGITALOCEAN_ACCESS_TOKEN }}
- name: Generate GITHUB_SHA
  id: github-sha
  shell: bash
  run: |
    SHORT_SHA=$(echo $GITHUB_SHA | cut -c1-7)
    echo "::set-output name=sha::$SHORT_SHA"
- name: Publish Image to Digital Ocean Container Registry
  shell: bash
  run: |
    doctl registry login --expiry-seconds 600
    docker build . -t registry.digitalocean.com/sample-go/add_sample:${{steps.github-sha.outputs.sha}}
    docker push registry.digitalocean.com/sample-go/add_sample:${{steps.github-sha.outputs.sha}}
   
as an example for short sha and push to docker??????

jonfriesen commented on README.md:
```diff
@@ -38,11 +44,12 @@ This action can be used to redeploy application on the DigitalOcean's [App Platf
 # Inputs
 - `app_name` - Name of the app on App Platform.
-- `list_of_image` - (optional)List of json object for providing information about name,repository and tag of the image in docr.(By default tag of the image is latest)
+- `list_of_image` - (optional)List of json object for providing information about name, repository and tag of the image in docr.(by default latest tag is used)
```
can we rename this to `images`?

jonfriesen replied:
also we advise against using `latest` should we use the commit short sha by default?

ParamPatel207 replied:
I believe user should have option to declare their own tags. For example; they may prefer to tag images by version number which might be unique.
Saying that we can provide an optional feature to deploy using sha tag but as users are deploying their docker images to docr by themselves we might need to provide handling of deployment from our side. For example; We take same input except tag field and then deploy dockerfile to docr with generated sha tag and then update our app platform app spec.

jonfriesen replied:
I'm suggesting we change the default from `latest` which has known issues to something that doesn't and is common across all projects. They're still welcome to set whatever they want but we can give them a nice default that is going to work without issue.

ParamPatel207 replied:
Actually they are the one pushing the images and we are the one pulling so what i am trying to say is they need to push image using sha hash. The default is latest because if the user does not mention tag DOCR uses latest as tag. If we set sha hash as default it might confuse users as now they have to explicitly set sha key and also push image using same key. This might be good in regulating the flow but I believe there may be many users who are not using tags for docr upload. So for example; I am user writing github action and I forgot to tag the image our action will be trying to retrieve tag based on hash rather then using latest.

jonfriesen replied:
Aaah I see, you're right. I misunderstood! thanks for explaining 👍

jonfriesen commented on internal/doctl/mylib.go:
```diff
@@ -0,0 +1,182 @@
+package mylib
```
change the package name to `doctl` and rename this file to `doctl.go`?

ParamPatel207 replied:
done

jonfriesen commented on internal/doctl/mylib.go:
```diff
@@ -0,0 +1,182 @@
+)
+
+//DoctlDependencies interface for doctl dependent functions
+type DoctlDependencies interface {
```
interfaces should be implemented where they are used. This interface should be in the `main.go`.

ParamPatel207 replied:
struct was not able to import interface from main

jonfriesen commented on internal/doctl/mylib.go:
```diff
@@ -0,0 +1,182 @@
+
+//DoctlServices is a struct for holding doctl dependent function interface
+type DoctlServices struct {
+	Dep DoctlDependencies
```
This probably shouldn't be in here.

ParamPatel207 replied:
Can you help me or provide suggestion for cleaning this up. I used this format for mock function creation. But ended up not using mock testing. Should I create Empty struct for DoctlServices?

jonfriesen commented on internal/doctl/mylib.go:
```diff
@@ -0,0 +1,182 @@
+}
+
+//IsAuthenticated used for user authentication
+func (d *DoctlServices) IsAuthenticated(token string) error {
```
```suggestion
func NewDoctlClient(token string) (*DoctlServices, error) {
```
Prefixing functions with `Is` typically means it checks if something is authenticated or not. This function appears to authenticate.

I think a good idea would be to change it into the client creation function, so it will try to authenticate and if it succeeds return a "cient" that you can call the other functions on.

ParamPatel207 replied:
updated to NewDoctlClient(token string) (*DoctlServices, error) and used the function for creation of DoctlServices ptr.

jonfriesen commented on internal/doctl/mylib.go:
```diff
@@ -0,0 +1,182 @@
+}
+
+//GetCurrentDeployment takes appID as input and returns list of deployments (used to retrieve most recent deployment)
+func (d *DoctlServices) GetCurrentDeployment(appID string) ([]byte, error) {
```
The command lists deployments other than the curernt deployment. The name can be updated to something more accurate like `ListDeployments` or the functionality can return the current deployment only.

ParamPatel207 replied:
updated using ListDeployments

jonfriesen commented on internal/doctl/mylib.go:
```diff
@@ -0,0 +1,182 @@
+	if err != nil {
+		return nil, errors.Wrap(err, "error in retrieving list of deployments")
+	}
+	return spec, nil
```
We could marshal this to a deployment list instead of a []byte right?

ParamPatel207 replied:
now function uses func (d *DoctlServices) ListDeployments(appID string) ([]godo.Deployment, error) {

jonfriesen commented on main.go:
```diff
@@ -6,225 +6,166 @@ import (
-	cmd := exec.Command("sh", "-c", fmt.Sprintf("doctl app create-deployment %s", appID))
-	_, err := cmd.Output()
+	//write to local temp file
+	err = ioutil.WriteFile(".do._app.yaml", newYaml, 0644)
```
Since we need to write the app spec to disk, can we write it to the GH Action temp dir? That way it's cleaned up between steps.

ParamPatel207 replied:
used https://pkg.go.dev/io/ioutil#TempFile

ParamPatel207 replied:
I was wondering if there is any GH action specific temp dir? Right now I updated tmp file creation using https://pkg.go.dev/io/ioutil#TempFile which will create unique tmp file for our use case.

ParamPatel207 committed:
Update README.md

update using suggestion

Co-authored-by: Jon Friesen <jfriesen@digitalocean.com>

ParamPatel207 committed:
Update README.md

based on suggestion

Co-authored-by: Jon Friesen <jfriesen@digitalocean.com>

ParamPatel207 committed:
Update README.md

update according to suggestion

Co-authored-by: Jon Friesen <jfriesen@digitalocean.com>

ParamPatel207 committed:
Update README.md

update according to suggestion

Co-authored-by: Jon Friesen <jfriesen@digitalocean.com>

ParamPatel207 committed:
Update README.md

Co-authored-by: Jon Friesen <jfriesen@digitalocean.com>

ParamPatel207 committed:
updated everything except readme

ParamPatel207 committed:
updated everything except readme

ParamPatel207 committed:
list_of_images -> images in action.yml

Commit:
Adds depedency injection, mocks

This change includes proper dependency injection and mock generation
for the doctl tool.

param-do committed:
Merge pull request #4 from ParamPatel207/jonfriesen/go_attempt

Adds depedency injection, mocks

ParamPatel207 committed:
updated mock test

ParamPatel207 committed:
updated test functions with mock and also updated the []byte consisteny problem. Moreover created parser folder in internal to maintain consistency and readability

ParamPatel207 committed:
Merge pull request #5 from ParamPatel207/go_attempt_mock

Go attempt mock

ParamPatel207 commented:
updated with refactoring of the code.
